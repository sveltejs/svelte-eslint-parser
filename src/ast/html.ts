import type ESTree from "estree";
import type { TSESTree } from "@typescript-eslint/types";
import type { BaseNode } from "./base.js";
import type { Token, Comment } from "./common.js";

export type SvelteHTMLNode =
  | SvelteProgram
  | SvelteScriptElement
  | SvelteStyleElement
  | SvelteElement
  | SvelteStartTag
  | SvelteEndTag
  | SvelteName
  | SvelteMemberExpressionName
  | SvelteText
  | SvelteLiteral
  | SvelteMustacheTag
  | SvelteDebugTag
  | SvelteConstTag
  | SvelteRenderTag
  | SvelteIfBlock
  | SvelteElseBlock
  | SvelteEachBlock
  | SvelteAwaitBlock
  | SvelteAwaitPendingBlock
  | SvelteAwaitThenBlock
  | SvelteAwaitCatchBlock
  | SvelteKeyBlock
  | SvelteSnippetBlock
  | SvelteAttribute
  | SvelteShorthandAttribute
  | SvelteSpreadAttribute
  | SvelteDirective
  | SvelteStyleDirective
  | SvelteSpecialDirective
  | SvelteGenericsDirective
  | SvelteDirectiveKey
  | SvelteSpecialDirectiveKey
  | SvelteHTMLComment;

/** Node of Svelte program root */
export interface SvelteProgram extends BaseNode {
  type: "Program";
  body: (SvelteScriptElement | SvelteStyleElement | Child)[];
  sourceType: "script" | "module";
  comments: Comment[];
  tokens: Token[];
  parent: null;
}

/** Node of elements like HTML element. */
export type SvelteElement =
  | SvelteHTMLElement
  | SvelteComponentElement
  | SvelteSpecialElement;

type BaseSvelteElement = BaseNode;

/** Node of `<script>` element. */
export interface SvelteScriptElement extends BaseSvelteElement {
  type: "SvelteScriptElement";
  name: SvelteName;
  startTag: SvelteStartTag;
  body: ESTree.Program["body"];
  endTag: SvelteEndTag | null;
  parent: SvelteProgram;
}
/** Node of `<style>` element. */
export interface SvelteStyleElement extends BaseSvelteElement {
  type: "SvelteStyleElement";
  name: SvelteName;
  startTag: SvelteStartTag;
  children: [SvelteText];
  endTag: SvelteEndTag | null;
  parent: SvelteProgram;
}
/** Node of HTML element. */
export interface SvelteHTMLElement extends BaseSvelteElement {
  type: "SvelteElement";
  kind: "html";
  name: SvelteName;
  startTag: SvelteStartTag;
  children: Child[];
  endTag: SvelteEndTag | null;
  parent:
    | SvelteProgram
    | SvelteElement
    | SvelteIfBlock
    | SvelteElseBlockAlone
    | SvelteEachBlock
    | SvelteAwaitPendingBlock
    | SvelteAwaitThenBlock
    | SvelteAwaitCatchBlock
    | SvelteKeyBlock
    | SvelteSnippetBlock;
}
/** Node of Svelte component element. */
export interface SvelteComponentElement extends BaseSvelteElement {
  type: "SvelteElement";
  kind: "component";
  name: ESTree.Identifier | SvelteMemberExpressionName;
  startTag: SvelteStartTag;
  children: Child[];
  endTag: SvelteEndTag | null;
  parent:
    | SvelteProgram
    | SvelteElement
    | SvelteIfBlock
    | SvelteElseBlockAlone
    | SvelteEachBlock
    | SvelteAwaitPendingBlock
    | SvelteAwaitThenBlock
    | SvelteAwaitCatchBlock
    | SvelteKeyBlock
    | SvelteSnippetBlock;
}
/** Node of Svelte special component element. e.g. `<svelte:window>` */
export interface SvelteSpecialElement extends BaseSvelteElement {
  type: "SvelteElement";
  kind: "special";
  name: SvelteName;
  startTag: SvelteStartTag;
  children: Child[];
  endTag: SvelteEndTag | null;
  parent:
    | SvelteProgram
    | SvelteElement
    | SvelteIfBlock
    | SvelteElseBlockAlone
    | SvelteEachBlock
    | SvelteAwaitPendingBlock
    | SvelteAwaitThenBlock
    | SvelteAwaitCatchBlock
    | SvelteKeyBlock
    | SvelteSnippetBlock;
}
/** Node of start tag. */
export interface SvelteStartTag extends BaseNode {
  type: "SvelteStartTag";
  attributes: (
    | SvelteAttribute
    | SvelteShorthandAttribute
    | SvelteSpreadAttribute
    | SvelteDirective
    | SvelteStyleDirective
    | SvelteSpecialDirective
    | SvelteGenericsDirective
  )[];
  selfClosing: boolean;
  parent: SvelteElement | SvelteScriptElement | SvelteStyleElement;
}
/** Node of end tag. */
export interface SvelteEndTag extends BaseNode {
  type: "SvelteEndTag";
  parent: SvelteElement | SvelteScriptElement | SvelteStyleElement;
}

/** Node of names. It is used for element names other than components and normal attribute names. */
export interface SvelteName extends BaseNode {
  type: "SvelteName";
  name: string;
  parent:
    | SvelteElement
    | SvelteScriptElement
    | SvelteStyleElement
    | SvelteAttribute
    | SvelteMemberExpressionName
    | SvelteDirectiveKey;
}

/** Nodes that may be used in component names. The component names separated by dots. */
export interface SvelteMemberExpressionName extends BaseNode {
  type: "SvelteMemberExpressionName";
  object: SvelteMemberExpressionName | ESTree.Identifier;
  property: SvelteName;
  parent: SvelteComponentElement;
}

type Child =
  | SvelteElement
  | SvelteText
  | SvelteMustacheTag
  | SvelteDebugTag
  | SvelteConstTag
  | SvelteRenderTag
  | SvelteIfBlockAlone
  | SvelteEachBlock
  | SvelteAwaitBlock
  | SvelteKeyBlock
  | SvelteSnippetBlock
  | SvelteHTMLComment;

/** Node of text. like HTML text. */
export interface SvelteText extends BaseNode {
  type: "SvelteText";
  value: string;
  parent:
    | SvelteProgram
    | SvelteElement
    | SvelteStyleElement
    | SvelteIfBlock
    | SvelteElseBlockAlone
    | SvelteEachBlock
    | SvelteAwaitPendingBlock
    | SvelteAwaitThenBlock
    | SvelteAwaitCatchBlock
    | SvelteKeyBlock
    | SvelteSnippetBlock;
}
/** Node of literal. */
export interface SvelteLiteral extends BaseNode {
  type: "SvelteLiteral";
  value: string;
  parent: SvelteAttribute | SvelteStyleDirective;
}

/** Node of mustache tag. e.g. `{...}`, `{@html ...}`. Like JSXExpressionContainer */
export type SvelteMustacheTag = SvelteMustacheTagText | SvelteMustacheTagRaw;
interface BaseSvelteMustacheTag extends BaseNode {
  type: "SvelteMustacheTag";
  kind: "text" | "raw";
  expression: ESTree.Expression;
  parent:
    | SvelteProgram
    | SvelteElement
    | SvelteIfBlock
    | SvelteElseBlockAlone
    | SvelteEachBlock
    | SvelteAwaitPendingBlock
    | SvelteAwaitThenBlock
    | SvelteAwaitCatchBlock
    | SvelteKeyBlock
    | SvelteSnippetBlock
    | SvelteAttribute
    | SvelteStyleDirective;
}
/** Node of mustache tag. e.g. `{...}``. Like JSXExpressionContainer */
export interface SvelteMustacheTagText extends BaseSvelteMustacheTag {
  kind: "text";
}
/** Node of mustache tag. e.g. `{@html ...}`. Like JSXExpressionContainer */
export interface SvelteMustacheTagRaw extends BaseSvelteMustacheTag {
  kind: "raw";
}
/** Node of debug mustache tag. e.g. `{@debug}` */
export interface SvelteDebugTag extends BaseNode {
  type: "SvelteDebugTag";
  identifiers: ESTree.Identifier[];
  parent:
    | SvelteProgram
    | SvelteElement
    | SvelteIfBlock
    | SvelteElseBlockAlone
    | SvelteEachBlock
    | SvelteAwaitPendingBlock
    | SvelteAwaitThenBlock
    | SvelteAwaitCatchBlock
    | SvelteKeyBlock
    | SvelteSnippetBlock
    | SvelteAttribute;
}
/** Node of const tag. e.g. `{@const}` */
export interface SvelteConstTag extends BaseNode {
  type: "SvelteConstTag";
  declaration: ESTree.VariableDeclarator;
  parent:
    | SvelteProgram
    | SvelteElement
    | SvelteIfBlock
    | SvelteElseBlockAlone
    | SvelteEachBlock
    | SvelteAwaitPendingBlock
    | SvelteAwaitThenBlock
    | SvelteAwaitCatchBlock
    | SvelteKeyBlock
    | SvelteSnippetBlock
    | SvelteAttribute;
}
/** Node of render tag. e.g. `{@render}` */
export interface SvelteRenderTag extends BaseNode {
  type: "SvelteRenderTag";
  expression:
    | ESTree.SimpleCallExpression
    | (ESTree.ChainExpression & { expression: ESTree.SimpleCallExpression });
  parent:
    | SvelteProgram
    | SvelteElement
    | SvelteIfBlock
    | SvelteElseBlockAlone
    | SvelteEachBlock
    | SvelteAwaitPendingBlock
    | SvelteAwaitThenBlock
    | SvelteAwaitCatchBlock
    | SvelteKeyBlock
    | SvelteSnippetBlock;
}
/** Node of if block. e.g. `{#if}` */
export type SvelteIfBlock = SvelteIfBlockAlone | SvelteIfBlockElseIf;
interface BaseSvelteIfBlock extends BaseNode {
  type: "SvelteIfBlock";
  elseif: boolean;
  expression: ESTree.Expression;
  children: Child[];
  else: SvelteElseBlock | null;
  parent:
    | SvelteProgram
    | SvelteElement
    | SvelteIfBlock
    | SvelteElseBlock
    | SvelteEachBlock
    | SvelteAwaitPendingBlock
    | SvelteAwaitThenBlock
    | SvelteAwaitCatchBlock
    | SvelteKeyBlock
    | SvelteSnippetBlock;
}
/** Node of if block. e.g. `{#if}` */
export interface SvelteIfBlockAlone extends BaseSvelteIfBlock {
  elseif: false;
  parent: Exclude<BaseSvelteIfBlock["parent"], SvelteElseBlockElseIf>;
}
/** Node of if block. e.g. `{:else if}` */
export interface SvelteIfBlockElseIf extends BaseSvelteIfBlock {
  elseif: true;
  parent: SvelteElseBlockElseIf;
}

/** Node of else block. e.g. `{:else}` */
export type SvelteElseBlock = SvelteElseBlockAlone | SvelteElseBlockElseIf;
interface BaseSvelteElseBlock extends BaseNode {
  type: "SvelteElseBlock";
  elseif: boolean;
  children: (Child | SvelteIfBlockElseIf)[];
  parent: SvelteIfBlock | SvelteEachBlock;
}
/** Node of else block. e.g. `{:else}` */
export interface SvelteElseBlockAlone extends BaseSvelteElseBlock {
  elseif: false;
  children: Child[];
}
/** Node of else block. e.g. `{:else if ...}` */
export interface SvelteElseBlockElseIf extends BaseSvelteElseBlock {
  elseif: true;
  children: [SvelteIfBlockElseIf];
}
/** Node of each block. e.g. `{#each}` */
export interface SvelteEachBlock extends BaseNode {
  type: "SvelteEachBlock";
  expression: ESTree.Expression;
  context: ESTree.Pattern | null;
  index: ESTree.Identifier | null;
  key: ESTree.Expression | null;
  children: Child[];
  else: SvelteElseBlockAlone | null;
  parent:
    | SvelteProgram
    | SvelteElement
    | SvelteIfBlock
    | SvelteElseBlockAlone
    | SvelteEachBlock
    | SvelteAwaitPendingBlock
    | SvelteAwaitThenBlock
    | SvelteAwaitCatchBlock
    | SvelteKeyBlock
    | SvelteSnippetBlock;
}
/** Node of await block. e.g. `{#await}`, `{#await ... then ... }`, `{#await ... catch ... }` */
export type SvelteAwaitBlock =
  | SvelteAwaitBlockAwaitPending
  | SvelteAwaitBlockAwaitThen
  | SvelteAwaitBlockAwaitCatch;
interface BaseSvelteAwaitBlock extends BaseNode {
  type: "SvelteAwaitBlock";
  expression: ESTree.Expression;
  kind: "await" | "await-then" | "await-catch";
  pending: SvelteAwaitPendingBlock | null;
  then: SvelteAwaitThenBlock | null;
  catch: SvelteAwaitCatchBlock | null;
  parent:
    | SvelteProgram
    | SvelteElement
    | SvelteIfBlock
    | SvelteElseBlockAlone
    | SvelteEachBlock
    | SvelteAwaitPendingBlock
    | SvelteAwaitThenBlock
    | SvelteAwaitCatchBlock
    | SvelteKeyBlock
    | SvelteSnippetBlock;
}
/** Node of await block. e.g. `{#await}` */
export interface SvelteAwaitBlockAwaitPending extends BaseSvelteAwaitBlock {
  kind: "await";
  pending: SvelteAwaitPendingBlock;
  then: SvelteAwaitThenBlockAlone | null;
  catch: SvelteAwaitCatchBlockAlone | null;
}
/** Node of await block. e.g. `{#await ... then ... }` */
export interface SvelteAwaitBlockAwaitThen extends BaseSvelteAwaitBlock {
  kind: "await-then";
  pending: null;
  then: SvelteAwaitThenBlockAwaitThen;
  catch: SvelteAwaitCatchBlockAlone | null;
}
/** Node of await block. e.g. `{#await ... catch ... }` */
export interface SvelteAwaitBlockAwaitCatch extends BaseSvelteAwaitBlock {
  kind: "await-catch";
  pending: null;
  then: null;
  catch: SvelteAwaitCatchBlockAwaitCatch;
}

/** Node of await pending block. e.g. `{#await expr} ... {:then}` */
export interface SvelteAwaitPendingBlock extends BaseNode {
  type: "SvelteAwaitPendingBlock";
  children: Child[];
  parent: SvelteAwaitBlock;
}
/** Node of await then block. e.g. `{:then}`, `{#await ... then ...}` */
export type SvelteAwaitThenBlock =
  | SvelteAwaitThenBlockAlone
  | SvelteAwaitThenBlockAwaitThen;
interface BaseSvelteAwaitThenBlock extends BaseNode {
  type: "SvelteAwaitThenBlock";
  awaitThen: boolean;
  value: ESTree.Pattern | null;
  children: Child[];
  parent: SvelteAwaitBlock;
}
/** Node of await then block. e.g. `{:then}` */
export interface SvelteAwaitThenBlockAlone extends BaseSvelteAwaitThenBlock {
  awaitThen: false;
  parent: SvelteAwaitBlockAwaitPending;
}
/** Node of await then block. e.g. `{#await ... then ...}` */
export interface SvelteAwaitThenBlockAwaitThen
  extends BaseSvelteAwaitThenBlock {
  awaitThen: true;
  parent: SvelteAwaitBlockAwaitThen;
}

/** Node of await catch block. e.g. `{:catch}`, `{#await ... catch ... }` */
export type SvelteAwaitCatchBlock =
  | SvelteAwaitCatchBlockAlone
  | SvelteAwaitCatchBlockAwaitCatch;
interface BaseSvelteAwaitCatchBlock extends BaseNode {
  type: "SvelteAwaitCatchBlock";
  awaitCatch: boolean;
  error: ESTree.Pattern | null;
  children: Child[];
  parent: SvelteAwaitBlock;
}
/** Node of await catch block. e.g. `{:catch}` */
export interface SvelteAwaitCatchBlockAlone extends BaseSvelteAwaitCatchBlock {
  awaitCatch: false;
  parent: SvelteAwaitBlockAwaitPending | SvelteAwaitBlockAwaitThen;
}
/** Node of await catch block. e.g. `{#await ... catch ... }` */
export interface SvelteAwaitCatchBlockAwaitCatch
  extends BaseSvelteAwaitCatchBlock {
  awaitCatch: true;
  error: ESTree.Pattern | null;
  children: Child[];
  parent: SvelteAwaitBlockAwaitCatch;
}
/** Node of key block. e.g. `{#key}` */
export interface SvelteKeyBlock extends BaseNode {
  type: "SvelteKeyBlock";
  expression: ESTree.Expression;
  children: Child[];
  parent:
    | SvelteProgram
    | SvelteElement
    | SvelteIfBlock
    | SvelteElseBlockAlone
    | SvelteEachBlock
    | SvelteAwaitPendingBlock
    | SvelteAwaitThenBlock
    | SvelteAwaitCatchBlock
    | SvelteKeyBlock
    | SvelteSnippetBlock;
}
/** Node of snippet block. e.g. `{#snippet}` */
export interface SvelteSnippetBlock extends BaseNode {
  type: "SvelteSnippetBlock";
  id: ESTree.Identifier;
  params: ESTree.Pattern[];
  children: Child[];
  parent:
    | SvelteProgram
    | SvelteElement
    | SvelteIfBlock
    | SvelteElseBlockAlone
    | SvelteEachBlock
    | SvelteAwaitPendingBlock
    | SvelteAwaitThenBlock
    | SvelteAwaitCatchBlock
    | SvelteKeyBlock
    | SvelteSnippetBlock;
}
/** Node of HTML comment. */
export interface SvelteHTMLComment extends BaseNode {
  type: "SvelteHTMLComment";
  value: string;
  parent:
    | SvelteProgram
    | SvelteElement
    | SvelteIfBlock
    | SvelteElseBlockAlone
    | SvelteEachBlock
    | SvelteAwaitPendingBlock
    | SvelteAwaitThenBlock
    | SvelteAwaitCatchBlock
    | SvelteKeyBlock
    | SvelteSnippetBlock;
}
/** Node of HTML attribute. */
export interface SvelteAttribute extends BaseNode {
  type: "SvelteAttribute";
  key: SvelteName;
  boolean: boolean;
  value: (SvelteLiteral | SvelteMustacheTagText)[];
  parent: SvelteStartTag;
}
/** Node of shorthand attribute. e.g. `<img {src}>` */
export interface SvelteShorthandAttribute extends BaseNode {
  type: "SvelteShorthandAttribute";
  key: ESTree.Identifier;
  value: ESTree.Identifier;
  parent: SvelteStartTag;
}
/** Node of spread attribute. e.g. `<Info {...pkg}/>`. Like JSXSpreadAttribute */
export interface SvelteSpreadAttribute extends BaseNode {
  type: "SvelteSpreadAttribute";
  argument: ESTree.Expression;
  parent: SvelteStartTag;
}

/** Node of directive. e.g. `<input bind:value />` */
export type SvelteDirective =
  | SvelteActionDirective
  | SvelteAnimationDirective
  | SvelteBindingDirective
  | SvelteClassDirective
  | SvelteEventHandlerDirective
  | SvelteLetDirective
  | SvelteRefDirective
  | SvelteTransitionDirective;

export type SvelteDirectiveKey =
  | SvelteDirectiveKeyTextName
  | SvelteDirectiveKeyFunctionName
  | SvelteDirectiveKeyForEventHandler
  | SvelteDirectiveKeyForAction
  | SvelteDirectiveKeyForStyleShorthand;
interface BaseSvelteDirectiveKey<N extends ESTree.Expression | SvelteName>
  extends BaseNode {
  type: "SvelteDirectiveKey";
  name: N;
  modifiers: string[];
  parent: SvelteDirective | SvelteStyleDirective;
}
export type SvelteDirectiveKeyTextName = BaseSvelteDirectiveKey<SvelteName>;
export type SvelteDirectiveKeyFunctionName =
  BaseSvelteDirectiveKey<ESTree.Identifier>;
export type SvelteDirectiveKeyForEventHandler =
  BaseSvelteDirectiveKey<SvelteName>;
export type SvelteDirectiveKeyForAction = BaseSvelteDirectiveKey<
  ESTree.Identifier | ESTree.MemberExpression | SvelteName
>;
export type SvelteDirectiveKeyForStyleShorthand =
  BaseSvelteDirectiveKey<ESTree.Identifier>;

interface BaseSvelteDirective extends BaseNode {
  type: "SvelteDirective";
  key: SvelteDirectiveKey;
  parent: SvelteStartTag;
}

export interface SvelteActionDirective extends BaseSvelteDirective {
  kind: "Action";
  key: SvelteDirectiveKeyForAction;
  expression: null | ESTree.Expression;
}
export interface SvelteAnimationDirective extends BaseSvelteDirective {
  kind: "Animation";
  key: SvelteDirectiveKeyFunctionName;
  expression: null | ESTree.Expression;
}
export interface SvelteBindingDirective extends BaseSvelteDirective {
  kind: "Binding";
  key: SvelteDirectiveKeyTextName;
  shorthand: boolean;
  expression: null | ESTree.Expression;
}
export interface SvelteClassDirective extends BaseSvelteDirective {
  kind: "Class";
  key: SvelteDirectiveKeyTextName;
  shorthand: boolean;
  expression: null | ESTree.Expression;
}
export interface SvelteEventHandlerDirective extends BaseSvelteDirective {
  kind: "EventHandler";
  key: SvelteDirectiveKeyForEventHandler;
  expression: null | ESTree.Expression;
}
export interface SvelteLetDirective extends BaseSvelteDirective {
  kind: "Let";
  key: SvelteDirectiveKeyTextName;
  expression: null | ESTree.Pattern;
}
export interface SvelteRefDirective extends BaseSvelteDirective {
  kind: "Ref";
  key: SvelteDirectiveKeyTextName;
  expression: null | ESTree.Expression;
}
export interface SvelteTransitionDirective extends BaseSvelteDirective {
  kind: "Transition";
  key: SvelteDirectiveKeyFunctionName;
  intro: boolean;
  outro: boolean;
  expression: null | ESTree.Expression;
}

/** Node of style directive. e.g. `<input style:color />` */
export type SvelteStyleDirective =
  | SvelteStyleDirectiveShorthand
  | SvelteStyleDirectiveLongform;
interface BaseSvelteStyleDirective extends BaseNode {
  type: "SvelteStyleDirective";
  key: SvelteDirectiveKeyTextName | SvelteDirectiveKeyForStyleShorthand;
  value: (SvelteLiteral | SvelteMustacheTagText)[];
  parent: SvelteStartTag;
}
export interface SvelteStyleDirectiveShorthand
  extends BaseSvelteStyleDirective {
  key: SvelteDirectiveKeyForStyleShorthand;
  shorthand: true;
  value: [];
}
export interface SvelteStyleDirectiveLongform extends BaseSvelteStyleDirective {
  key: SvelteDirectiveKeyTextName;
  shorthand: false;
  value: (SvelteLiteral | SvelteMustacheTagText)[];
}
export interface SvelteSpecialDirectiveKey extends BaseNode {
  type: "SvelteSpecialDirectiveKey";
  parent: SvelteSpecialDirective;
}
export interface SvelteSpecialDirective extends BaseNode {
  type: "SvelteSpecialDirective";
  kind: "this";
  key: SvelteSpecialDirectiveKey;
  expression: ESTree.Expression;
  parent: SvelteStartTag /* & { parent: SvelteSpecialElement } */;
}
export interface SvelteGenericsDirective extends BaseNode {
  type: "SvelteGenericsDirective";
  key: SvelteName & { name: "generics" };
  params: TSESTree.TSTypeParameterDeclaration["params"];
  parent: SvelteStartTag /* & { parent: SvelteScriptElement } */;
}
