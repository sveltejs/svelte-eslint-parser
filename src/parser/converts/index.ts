export { convertSvelteRoot } from "./root.js";
