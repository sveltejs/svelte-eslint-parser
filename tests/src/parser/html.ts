import * as chai from "chai";
import { jestSnapshotPlugin } from "mocha-chai-jest-snapshot";

import { parseAttributes } from "../../../src/parser/html";

chai.use(jestSnapshotPlugin());
describe("parseAttributes", () => {
  const testCases = [
    {
      input: 'attr="value"',
    },
    {
      input: '<script lang="ts">',
      index: 7,
    },
    {
      input: "<script lang='ts'>",
      index: 7,
    },
    {
      input: "<script lang=ts>",
      index: 7,
    },
    {
      input: "<style global>",
      index: 6,
    },
    {
      input: "<style global/>",
      index: 6,
    },
    {
      input: "",
    },
    {
      input: 'attr  =  "value"',
    },
    {
      input: "attr",
    },
    {
      input: "attr  ",
    },
    {
      input: `empty=""  `,
    },
    {
      input: `empty=''  `,
    },
    {
      input: `quote="'"  `,
    },
    {
      input: `quote='"'  `,
    },
    {
      input: `expr={true}  `,
    },
    {
      input: `expr="{true}"  `,
    },
    {
      input: `expr='{true}'  `,
    },
    {
      input: `expr={"s"}  `,
    },
    {
      input: `expr={"}"}  `,
    },
    {
      input: `expr={/*}*/"}"}  `,
    },
    {
      input: `expr={/*}*///}\n"}"}  `,
    },
  ];
  for (const { input, index } of testCases) {
    it(input || "(empty)", () => {
      chai.expect(parseAttributes(input, index || 0)).toMatchSnapshot();
    });
  }
});
