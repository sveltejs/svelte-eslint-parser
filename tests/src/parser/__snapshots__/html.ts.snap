// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseAttributes (empty) 1`] = `
Object {
  "attributes": Array [],
  "index": 0,
}
`;

exports[`parseAttributes <script lang="ts"> 1`] = `
Object {
  "attributes": Array [
    Object {
      "end": 17,
      "name": "lang",
      "start": 8,
      "type": "Attribute",
      "value": Array [
        Object {
          "data": "ts",
          "end": 16,
          "raw": "ts",
          "start": 14,
          "type": "Text",
        },
      ],
    },
  ],
  "index": 17,
}
`;

exports[`parseAttributes <script lang='ts'> 1`] = `
Object {
  "attributes": Array [
    Object {
      "end": 17,
      "name": "lang",
      "start": 8,
      "type": "Attribute",
      "value": Array [
        Object {
          "data": "ts",
          "end": 16,
          "raw": "ts",
          "start": 14,
          "type": "Text",
        },
      ],
    },
  ],
  "index": 17,
}
`;

exports[`parseAttributes <script lang=ts> 1`] = `
Object {
  "attributes": Array [
    Object {
      "end": 15,
      "name": "lang",
      "start": 8,
      "type": "Attribute",
      "value": Array [
        Object {
          "data": "ts",
          "end": 15,
          "raw": "ts",
          "start": 13,
          "type": "Text",
        },
      ],
    },
  ],
  "index": 15,
}
`;

exports[`parseAttributes <style global/> 1`] = `
Object {
  "attributes": Array [
    Object {
      "end": 13,
      "name": "global",
      "start": 7,
      "type": "Attribute",
      "value": true,
    },
  ],
  "index": 13,
}
`;

exports[`parseAttributes <style global> 1`] = `
Object {
  "attributes": Array [
    Object {
      "end": 13,
      "name": "global",
      "start": 7,
      "type": "Attribute",
      "value": true,
    },
  ],
  "index": 13,
}
`;

exports[`parseAttributes attr   1`] = `
Object {
  "attributes": Array [
    Object {
      "end": 4,
      "name": "attr",
      "start": 0,
      "type": "Attribute",
      "value": true,
    },
  ],
  "index": 6,
}
`;

exports[`parseAttributes attr  =  "value" 1`] = `
Object {
  "attributes": Array [
    Object {
      "end": 16,
      "name": "attr",
      "start": 0,
      "type": "Attribute",
      "value": Array [
        Object {
          "data": "value",
          "end": 15,
          "raw": "value",
          "start": 10,
          "type": "Text",
        },
      ],
    },
  ],
  "index": 16,
}
`;

exports[`parseAttributes attr 1`] = `
Object {
  "attributes": Array [
    Object {
      "end": 4,
      "name": "attr",
      "start": 0,
      "type": "Attribute",
      "value": true,
    },
  ],
  "index": 4,
}
`;

exports[`parseAttributes attr="value" 1`] = `
Object {
  "attributes": Array [
    Object {
      "end": 12,
      "name": "attr",
      "start": 0,
      "type": "Attribute",
      "value": Array [
        Object {
          "data": "value",
          "end": 11,
          "raw": "value",
          "start": 6,
          "type": "Text",
        },
      ],
    },
  ],
  "index": 12,
}
`;

exports[`parseAttributes empty=""   1`] = `
Object {
  "attributes": Array [
    Object {
      "end": 8,
      "name": "empty",
      "start": 0,
      "type": "Attribute",
      "value": Array [
        Object {
          "data": "",
          "end": 7,
          "raw": "",
          "start": 7,
          "type": "Text",
        },
      ],
    },
  ],
  "index": 10,
}
`;

exports[`parseAttributes empty=''   1`] = `
Object {
  "attributes": Array [
    Object {
      "end": 8,
      "name": "empty",
      "start": 0,
      "type": "Attribute",
      "value": Array [
        Object {
          "data": "",
          "end": 7,
          "raw": "",
          "start": 7,
          "type": "Text",
        },
      ],
    },
  ],
  "index": 10,
}
`;

exports[`parseAttributes expr="{true}"   1`] = `
Object {
  "attributes": Array [
    Object {
      "end": 13,
      "name": "expr",
      "start": 0,
      "type": "Attribute",
      "value": Array [
        Object {
          "end": 13,
          "expression": Object {
            "end": 11,
            "leadingComments": Array [],
            "raw": "true",
            "start": 7,
            "type": "Literal",
            "value": true,
          },
          "start": 5,
          "type": "ExpressionTag",
        },
      ],
    },
  ],
  "index": 15,
}
`;

exports[`parseAttributes expr='{true}'   1`] = `
Object {
  "attributes": Array [
    Object {
      "end": 13,
      "name": "expr",
      "start": 0,
      "type": "Attribute",
      "value": Array [
        Object {
          "end": 13,
          "expression": Object {
            "end": 11,
            "leadingComments": Array [],
            "raw": "true",
            "start": 7,
            "type": "Literal",
            "value": true,
          },
          "start": 5,
          "type": "ExpressionTag",
        },
      ],
    },
  ],
  "index": 15,
}
`;

exports[`parseAttributes expr={"}"}   1`] = `
Object {
  "attributes": Array [
    Object {
      "end": 10,
      "name": "expr",
      "start": 0,
      "type": "Attribute",
      "value": Array [
        Object {
          "end": 10,
          "expression": Object {
            "end": 9,
            "leadingComments": Array [],
            "raw": "\\"}\\"",
            "start": 6,
            "type": "Literal",
            "value": "}",
          },
          "start": 5,
          "type": "ExpressionTag",
        },
      ],
    },
  ],
  "index": 12,
}
`;

exports[`parseAttributes expr={"s"}   1`] = `
Object {
  "attributes": Array [
    Object {
      "end": 10,
      "name": "expr",
      "start": 0,
      "type": "Attribute",
      "value": Array [
        Object {
          "end": 10,
          "expression": Object {
            "end": 9,
            "leadingComments": Array [],
            "raw": "\\"s\\"",
            "start": 6,
            "type": "Literal",
            "value": "s",
          },
          "start": 5,
          "type": "ExpressionTag",
        },
      ],
    },
  ],
  "index": 12,
}
`;

exports[`parseAttributes expr={/*}*/"}"}   1`] = `
Object {
  "attributes": Array [
    Object {
      "end": 15,
      "name": "expr",
      "start": 0,
      "type": "Attribute",
      "value": Array [
        Object {
          "end": 15,
          "expression": Object {
            "end": 14,
            "leadingComments": Array [
              Object {
                "end": 11,
                "start": 6,
                "type": "Block",
                "value": "}",
              },
            ],
            "raw": "\\"}\\"",
            "start": 11,
            "type": "Literal",
            "value": "}",
          },
          "start": 5,
          "type": "ExpressionTag",
        },
      ],
    },
  ],
  "index": 17,
}
`;

exports[`parseAttributes expr={/*}*///}
"}"}   1`] = `
Object {
  "attributes": Array [
    Object {
      "end": 19,
      "name": "expr",
      "start": 0,
      "type": "Attribute",
      "value": Array [
        Object {
          "end": 19,
          "expression": Object {
            "end": 18,
            "leadingComments": Array [
              Object {
                "end": 11,
                "start": 6,
                "type": "Block",
                "value": "}",
              },
              Object {
                "end": 15,
                "start": 11,
                "type": "Line",
                "value": "}",
              },
            ],
            "raw": "\\"}\\"",
            "start": 15,
            "type": "Literal",
            "value": "}",
          },
          "start": 5,
          "type": "ExpressionTag",
        },
      ],
    },
  ],
  "index": 21,
}
`;

exports[`parseAttributes expr={true}   1`] = `
Object {
  "attributes": Array [
    Object {
      "end": 11,
      "name": "expr",
      "start": 0,
      "type": "Attribute",
      "value": Array [
        Object {
          "end": 11,
          "expression": Object {
            "end": 10,
            "leadingComments": Array [],
            "raw": "true",
            "start": 6,
            "type": "Literal",
            "value": true,
          },
          "start": 5,
          "type": "ExpressionTag",
        },
      ],
    },
  ],
  "index": 13,
}
`;

exports[`parseAttributes quote="'"   1`] = `
Object {
  "attributes": Array [
    Object {
      "end": 9,
      "name": "quote",
      "start": 0,
      "type": "Attribute",
      "value": Array [
        Object {
          "data": "'",
          "end": 8,
          "raw": "'",
          "start": 7,
          "type": "Text",
        },
      ],
    },
  ],
  "index": 11,
}
`;

exports[`parseAttributes quote='"'   1`] = `
Object {
  "attributes": Array [
    Object {
      "end": 9,
      "name": "quote",
      "start": 0,
      "type": "Attribute",
      "value": Array [
        Object {
          "data": "\\"",
          "end": 8,
          "raw": "\\"",
          "start": 7,
          "type": "Text",
        },
      ],
    },
  ],
  "index": 11,
}
`;
